core/domain/topic_validators.py

```diff
@@ -0,0 +1,585 @@
+                topic_models.TopicSnapshotMetadataModel, snapshot_model_ids),
+            base_model_validators.ExternalModelFetcherDetails(
+                'snapshot_content_ids', topic_models.TopicSnapshotContentModel,
+                snapshot_model_ids),
+            base_model_validators.ExternalModelFetcherDetails(
+                'story_ids', story_models.StoryModel,
+                canonical_story_ids + additional_story_ids),
+            base_model_validators.ExternalModelFetcherDetails(
+                'skill_ids', skill_models.SkillModel, skill_ids),
+            base_model_validators.ExternalModelFetcherDetails(
+                'subtopic_page_ids',
+                subtopic_models.SubtopicPageModel,
+                ['%s-%s' % (
+                    item.id, subtopic['id']) for subtopic in item.subtopics])]
+
+    @classmethod
+    def _validate_canonical_name_is_unique(cls, item):
+        """Validate that canonical name of the model unique.
+
+        Args:
+            item: datastore_services.Model. TopicModel to validate.
+        """
+        topic_models_list = topic_models.TopicModel.query().filter(
+            topic_models.TopicModel.canonical_name == (
+                item.canonical_name)).filter(
+                    topic_models.TopicModel.deleted == False).fetch() # pylint: disable=singleton-comparison
+        topic_model_ids = [
+            topic_model.id
+            for topic_model in topic_models_list if topic_model.id != item.id]
+        if topic_model_ids:
+            cls._add_error(
+                'unique %s' % base_model_validators.ERROR_CATEGORY_NAME_CHECK,
+                'Entity id %s: canonical name %s matches with canonical '
+                'name of topic models with ids %s' % (
+                    item.id, item.canonical_name, topic_model_ids))
+
+    @classmethod
+    def _validate_canonical_name_matches_name_in_lowercase(cls, item):
+        """Validate that canonical name of the model is same as name of the
+        model in lowercase.
+
+        Args:
+            item: datastore_services.Model. TopicModel to validate.
+        """
+        name = item.name
+        if name.lower() != item.canonical_name:
+            cls._add_error(
+                'canonical %s' % (
+                    base_model_validators.ERROR_CATEGORY_NAME_CHECK),
+                'Entity id %s: Entity name %s in lowercase does not match '
+                'canonical name %s' % (item.id, item.name, item.canonical_name))
+
+    @classmethod
+    def _validate_uncategorized_skill_ids_not_in_subtopic_skill_ids(cls, item):
+        """Validate that uncategorized_skill_ids of model is not present in
+        any subtopic of the model.
+
+        Args:
+            item: datastore_services.Model. TopicModel to validate.
+        """
+        for skill_id in item.uncategorized_skill_ids:
+            for subtopic in item.subtopics:
+                if skill_id in subtopic['skill_ids']:
```

Comment:
Can't this be done more efficiently? Since we are expecting a topic to have quite a lot of skills. Like having 2 separate lists for uncategorized skills and all skills in a subtopic and checking set difference or intersection maybe?

Reply:
Ditto as above*

Reply:
@aks681 I believe this might best be handled as another separate issue, or would it best be included in the issue created for the above comment?

These seem sufficiently different to me to warrant two tickets.

Reply:
I think you can have it in the same issue since both are technically optimizations to topic_validators, so having 2 issues might be harder to track.

Reply:
@aks681, We have already created an issue at #11085, PTAL! (I've also left a comment for you in the issue thread)